### PR TITLE
Cancel rsvp

### LINKED ISSUE #474: cmd + z  RSVP, if accidentally rsvp'd

### PR DESCRIPTION
**Pull request by TraceyOnim**
closes #474 

Peek 2021-08-31 12-32

### CODE REVIEW

File: lib/quadquizaminos/contests.ex [hunk 1]
```diff
@@ -244,6 +244,18 @@ defmodule Quadquizaminos.Contests do
...
+  @doc """
+  deletes RSVP from the database
+  """
+  @spec cancel_rsvp(integer(), User.t()) :: {:ok, RSVP.t()} | {:error, Changeset.t()}
+  def cancel_rsvp(contest_id, %User{} = user) do
+    contest = get_contest(contest_id)
+
+    user
+    |> RSVP.user_contest_rsvp_query(contest)
+    |> Repo.delete_all()
```
Comment (sigu) [inline lib/quadquizaminos/contests.ex:256]: We expect only one RSVP per user so `Repo.delete` can work in this case

Comment (TraceyOnim) [reply lib/quadquizaminos/contests.ex]: I have rectified it

File: lib/quadquizaminos/contests.ex [hunk 2]
```diff
@@ -244,6 +244,25 @@ defmodule Quadquizaminos.Contests do
...
+  deletes RSVP from the database
+  """
+  @spec cancel_rsvp(integer(), User.t()) :: {:ok, RSVP.t()} | {:error, Changeset.t()}
+  def cancel_rsvp(contest_id, %User{} = user) do
+    contest_id
+    |> get_contest()
+    |> _cancel_rsvp(user)
+  end
+
+  defp _cancel_rsvp(nil, _user), do: {0, []}
```
Comment (TraceyOnim) [inline lib/quadquizaminos/contests.ex:257]:
```suggestion
  defp _cancel_rsvp(nil, _user), do: {:error, changeset}
```

File: lib/quadquizaminos/contests.ex [hunk 3]
```diff
@@ -244,6 +244,25 @@ defmodule Quadquizaminos.Contests do
     RSVP.changeset(rsvp, attrs, %User{})
   end
 
+  @doc """
+  deletes RSVP from the database
+  """
+  @spec cancel_rsvp(integer(), User.t()) :: {:ok, RSVP.t()} | {:error, Changeset.t()}
+  def cancel_rsvp(contest_id, %User{} = user) do
+    contest_id
+    |> get_contest()
```
Comment (sigu) [inline lib/quadquizaminos/contests.ex:253]: Why do we need to get the contest first?

Comment (TraceyOnim) [reply lib/quadquizaminos/contests.ex:253]: I needed to use the contest struct.
`user_contest_rsvp_query/2` takes in the user struct and contest struct.

**Review by emmyachieng — APPROVED** [review]

**Review by WanjikuMac — APPROVED** [review]

File: test/quadquizaminos/contests/contest/rsvp_test.exs [hunk 4]
```diff
@@ -5,11 +5,29 @@ defmodule Quadquizaminos.ContestsTest.RSVP do
...
+      {:ok, current_user} = Accounts.create_user(%User{}, %{name: "ha", uid: "34"})
+      {:ok, contest} = Contests.create_contest(%{name: "test contest"})
+      attrs = %{"contest_id" => contest.id}
 
-    {:ok, %RSVP{}} = Contests.create_rsvp(attrs, current_user)
+      [attrs: attrs, current_user: current_user, contest: contest]
+    end
+
+    test "create_rsvp/2 creates a new rsvp on db", %{attrs: attrs, current_user: current_user} do
+      {:ok, %RSVP{}} = Contests.create_rsvp(attrs, current_user)
```
Comment (WanjikuMac) [inline test/quadquizaminos/contests/contest/rsvp_test.exs:18]: Don't we need to assert that a new rsvp has been created?

Comment (TraceyOnim) [reply test/quadquizaminos/contests/contest/rsvp_test.exs:18]: With `{: ok, rsvp}` also indicates that rsvp was successfully inserted into the db.

Comment (WanjikuMac) [reply test/quadquizaminos/contests/contest/rsvp_test.exs:18]: But what happens when the rsvp is not successfully written to the DB? I find assertions make it easy to outline what was expected versus what was returned, saving you debugging time

**Pull request merged**